pkg/services/ngalert/sender/router.go

```diff
@@ -0,0 +1,268 @@
+package sender
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/benbjohnson/clock"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/notifier"
+	"github.com/grafana/grafana/pkg/services/ngalert/store"
+)
+
+// AlertsRouter handles alerts generated during alert rule evaluation.
+// Based on rule's orgID and the configuration for that organization,
+// it determines whether an alert needs to be sent to an external Alertmanager and\or internal notifier.Alertmanager
+//
+// After creating a Dispatcher, you must call Run to keep the Dispatcher's
+// state synchronized with the alerting configuration.
+type AlertsRouter struct {
+	AdminConfigMtx   sync.RWMutex
+	logger           log.Logger
+	clock            clock.Clock
+	adminConfigStore store.AdminConfigurationStore
+
+	// Senders help us send alerts to external Alertmanagers.
+	Senders          map[int64]*Sender
+	SendersCfgHash   map[int64]string
+	MultiOrgNotifier *notifier.MultiOrgAlertmanager
+	SendAlertsTo     map[int64]models.AlertmanagersChoice
+
+	appURL                  *url.URL
+	disabledOrgs            map[int64]struct{}
+	adminConfigPollInterval time.Duration
+}
+
+func NewAlertsRouter(multiOrgNotifier *notifier.MultiOrgAlertmanager, store store.AdminConfigurationStore, clk clock.Clock, appURL *url.URL, disabledOrgs map[int64]struct{}, configPollInterval time.Duration) *AlertsRouter {
+	d := &AlertsRouter{
+		AdminConfigMtx:   sync.RWMutex{},
+		logger:           log.New("ngalert-notifications-dispatcher"),
+		clock:            clk,
+		adminConfigStore: store,
+
+		Senders:          map[int64]*Sender{},
+		SendersCfgHash:   map[int64]string{},
+		MultiOrgNotifier: multiOrgNotifier,
+		SendAlertsTo:     map[int64]models.AlertmanagersChoice{},
+
+		appURL:                  appURL,
+		disabledOrgs:            disabledOrgs,
+		adminConfigPollInterval: configPollInterval,
+	}
+	return d
+}
+
+func (d *AlertsRouter) adminConfigSync(ctx context.Context) error {
```

Comment:
I think we should inline this within the `run` function - I don't see us gaining anything from this level of indirection. It doesn't help with testing and it's [quite](https://github.com/prometheus/prometheus/blob/main/rules/manager.go#L341-L350) [common](https://github.com/grafana/mimir/blob/main/pkg/ruler/ruler.go#L428-L458) for control functions (e.g. `run`) to have this structure.

Reply:
Agree. Moved it.